core/example/benchmark/Cabana_CommPerformance.cpp

```diff
@@ -0,0 +1,586 @@
+        {
+            file << std::endl;
+            file << _name << std::endl;
+            file << "num_rank " << data_point_name << " min max ave" << std::endl;
+        }
+
+        // Write out each data point
+        for ( std::size_t n = 0; n < _data.size(); ++n )
+        {
+            if ( !_is_stopped[n] )
+                throw std::logic_error( "attempted to output from a running timer" );
+
+            // Compute the minimum.
+            double local_min = (double) std::min_element(
+                _data[n].begin(), _data[n].end() )->count();
+            double global_min = 0.0;
+            MPI_Reduce( &local_min, &global_min, 1, MPI_DOUBLE, MPI_MIN, 0, _comm );
+
+            // Compute the maximum.
+            double local_max = (double) std::max_element(
+                _data[n].begin(), _data[n].end() )->count();
+            double global_max = 0.0;
+            MPI_Reduce( &local_max, &global_max, 1, MPI_DOUBLE, MPI_MAX, 0, _comm );
+
+            // Compute the average.
+            double local_sum = 0.0;
+            for ( auto d : _data[n] )
+                local_sum += (double) d.count();
+            double average = 0.0;
+            MPI_Reduce( &local_sum, &average, 1, MPI_DOUBLE, MPI_SUM, 0, _comm );
+            average /= _data[n].size() * comm_size;
+
+            // Output on rank 0.
+            if ( 0 == comm_rank )
+            {
+                file << comm_size << " "
+                     << data_point_vals[n] << " "
+                     << global_min << " "
+                     << global_max << " "
+                     << average << std::endl;
+            }
+        }
+    }
+
+  private:
+
+    MPI_Comm _comm;
+    std::string _name;
+    std::vector<std::size_t> _byte_sizes;
+    std::vector<std::vector<std::chrono::high_resolution_clock::time_point> > _starts;
+    std::vector<std::vector<std::chrono::microseconds> > _data;
+    std::vector<bool> _is_stopped;
+};
+
+//---------------------------------------------------------------------------//
+// Performance test.
+// Data device type is where the data to be communicated lives.
+// Comm device type is the device we want to use for communication.
+template<class DataDevice, class CommDevice>
+void performanceTest( std::fstream& file,
+                      const std::size_t num_particle,
+                      const std::string& test_prefix )
+{
+    // PROBLEM SETUP
+    // -------------
+
+    // Get comm world.
+    auto comm = MPI_COMM_WORLD;
+
+    // Get comm size;
+    int comm_size;
+    MPI_Comm_size( comm, &comm_size );
+
+    // Partition the problem in 3 dimensions.
+    const int space_dim = 3;
+    std::vector<int> ranks_per_dim( space_dim );
+    MPI_Dims_create( comm_size, space_dim, ranks_per_dim.data() );
+
+    // Generate a communicator with a cartesian topology and periodic
+    // boundaries.
+    std::vector<int> periodic_dims( space_dim, 1 );
+    MPI_Comm cart_comm;
+    int reorder_cart_ranks = 1;
```

Comment:
I was thinking that it could possibly result in a more optimal layout of processors for the type of communication I am doing in this test (essentially Cartesian grid halo exchange). It would be interesting to turn this off though and see if/how performance changes.